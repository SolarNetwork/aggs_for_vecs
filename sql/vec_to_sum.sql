-- smallint sum lots
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint sum none
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = -1;
-- smallint sum one null
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = 1;
-- smallint sum array of nulls
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = 2;
-- smallint sum one not-null
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = 4;
-- smallint sum array of nulls and one other
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int sum lots
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int sum none
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = -1;
-- int sum one null
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = 1;
-- int sum array of nulls
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = 2;
-- int sum one not-null
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = 4;
-- int sum array of nulls and one other
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint sum lots
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint sum none
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = -1;
-- bigint sum one null
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = 1;
-- bigint sum array of nulls
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = 2;
-- bigint sum one not-null
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = 4;
-- bigint sum array of nulls and one other
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real sum lots
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real sum none
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = -1;
-- real sum one null
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = 1;
-- real sum array of nulls
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = 2;
-- real sum one not-null
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = 4;
-- real sum array of nulls and one other
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 sum lots
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float8 sum none
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = -1;
-- float8 sum one null
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 1;
-- float8 sum array of nulls
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 2;
-- float8 sum one not-null
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 4;
-- float8 sum array of nulls and one other
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id IN (2, 4);

-- numeric sum lots
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- numeric sum none
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = -1;
-- numeric sum one null
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = 1;
-- numeric sum array of nulls
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = 2;
-- numeric sum one not-null
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = 4;
-- numeric sum array of nulls and one other
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id IN (2, 4);
-- numeric sum data 01
-- results expected to match:
-- WITH di AS (
-- 	SELECT idx_i, sum(val_i) AS sum_i
-- 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
-- 	GROUP BY idx_i
-- )
-- SELECT array_agg(sum_i ORDER BY idx_i) AS data_i_sum FROM di
SELECT vec_to_sum(data_i) FROM measurements2;
