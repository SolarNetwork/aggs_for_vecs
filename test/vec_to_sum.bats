load test_helper

@test "int16 sum" {
  result="$(query "SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{2,2,5}" ]
}

# @test "int32 sum" {
  # result="$(query "SELECT vec_to_sum('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 sum" {
  # result="$(query "SELECT vec_to_sum('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 sum" {
  # result="$(query "SELECT vec_to_sum('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 sum lots" {
  result="$(query "SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{2,2,5}" ]
}

@test "float8 sum none" {
  result="$(query "SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 sum one null" {
  result="$(query "SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 sum array of nulls" {
  result="$(query "SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "float8 sum one not-null" {
  result="$(query "SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "float8 sum array of nulls and one other" {
  result="$(query "SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "string sum" {
  run query "SELECT vec_to_sum(vals) FROM (VALUES (ARRAY['a']), (ARRAY['b'])) t(vals)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  vec_to_sum input must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION" ]
}
