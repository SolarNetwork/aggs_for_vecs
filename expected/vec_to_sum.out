-- smallint sum lots
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_sum 
------------
 {2,2,5}
(1 row)

-- smallint sum none
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_sum 
------------
 
(1 row)

-- smallint sum one null
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_sum 
------------
 
(1 row)

-- smallint sum array of nulls
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = 2;
    vec_to_sum    
------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint sum one not-null
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- smallint sum array of nulls and one other
SELECT vec_to_sum(smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- int sum lots
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_sum 
------------
 {2,2,5}
(1 row)

-- int sum none
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = -1;
 vec_to_sum 
------------
 
(1 row)

-- int sum one null
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = 1;
 vec_to_sum 
------------
 
(1 row)

-- int sum array of nulls
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = 2;
    vec_to_sum    
------------------
 {NULL,NULL,NULL}
(1 row)

-- int sum one not-null
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id = 4;
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- int sum array of nulls and one other
SELECT vec_to_sum(ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- bigint sum lots
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_sum 
------------
 {2,2,5}
(1 row)

-- bigint sum none
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_sum 
------------
 
(1 row)

-- bigint sum one null
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_sum 
------------
 
(1 row)

-- bigint sum array of nulls
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = 2;
    vec_to_sum    
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint sum one not-null
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- bigint sum array of nulls and one other
SELECT vec_to_sum(bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- real sum lots
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_sum 
------------
 {2,2,5}
(1 row)

-- real sum none
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = -1;
 vec_to_sum 
------------
 
(1 row)

-- real sum one null
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = 1;
 vec_to_sum 
------------
 
(1 row)

-- real sum array of nulls
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = 2;
    vec_to_sum    
------------------
 {NULL,NULL,NULL}
(1 row)

-- real sum one not-null
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id = 4;
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- real sum array of nulls and one other
SELECT vec_to_sum(reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- float8 sum lots
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_sum 
------------
 {2,2,5}
(1 row)

-- float8 sum none
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = -1;
 vec_to_sum 
------------
 
(1 row)

-- float8 sum one null
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 1;
 vec_to_sum 
------------
 
(1 row)

-- float8 sum array of nulls
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 2;
    vec_to_sum    
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 sum one not-null
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id = 4;
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- float8 sum array of nulls and one other
SELECT vec_to_sum(floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_sum 
------------
 {1,NULL,2}
(1 row)

-- numeric sum lots
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
    vec_to_sum    
------------------
 {2.46,2.34,5.79}
(1 row)

-- numeric sum none
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = -1;
 vec_to_sum 
------------
 
(1 row)

-- numeric sum one null
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = 1;
 vec_to_sum 
------------
 
(1 row)

-- numeric sum array of nulls
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = 2;
    vec_to_sum    
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric sum one not-null
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id = 4;
    vec_to_sum    
------------------
 {1.23,NULL,2.34}
(1 row)

-- numeric sum array of nulls and one other
SELECT vec_to_sum(nums) FROM measurements WHERE sensor_id IN (2, 4);
    vec_to_sum    
------------------
 {1.23,NULL,2.34}
(1 row)

-- numeric sum data 01
-- results expected to match:
-- WITH di AS (
-- 	SELECT idx_i, sum(val_i) AS sum_i
-- 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
-- 	GROUP BY idx_i
-- )
-- SELECT array_agg(sum_i ORDER BY idx_i) AS data_i_sum FROM di
SELECT vec_to_sum(data_i) FROM measurements2;
                                             vec_to_sum                                              
-----------------------------------------------------------------------------------------------------
 {35968097,3460.738929,42273.93922,9198.984594,25360099,183.07443653,-971.9407094,46903731,25143722}
(1 row)

